File: r-package/R/answer_checker.R
# =============================================================================
# Answer Checker — Equivalence-aware answer checking with flexible matching
# =============================================================================

#' Normalize an answer string for comparison
#'
#' Strips outer parentheses, whitespace, normalizes case, and removes
#' common formatting artifacts. Applied to both student and expected answers
#' before comparison.
#'
#' @param answer Character string
#' @return Normalized character string
#' @keywords internal
normalize_answer <- function(answer) {
  answer <- trimws(as.character(answer))
  answer <- tolower(answer)
  # Strip outer parentheses: "(a)" -> "a", "(3/4)" -> "3/4"
  # Only strip if the content has no unmatched inner parens
  answer <- gsub("^\\(([^()]+)\\)$", "\\1", answer)
  # Normalize LaTeX artifacts that MathLive may produce
  answer <- gsub("\\\\dfrac", "\\\\frac", answer)   # \dfrac -> \frac
  answer <- gsub("\\\\left\\s*", "", answer)          # \left( -> (
  answer <- gsub("\\\\right\\s*", "", answer)         # \right) -> )
  answer <- gsub("\\\\cdot", "*", answer)             # \cdot -> *
  answer <- gsub("\\\\times", "*", answer)            # \times -> *
  answer <- gsub("\\\\,", "", answer)                 # thin space
  answer <- gsub("\\\\;", "", answer)                 # medium space
  answer <- gsub("\\\\!", "", answer)                 # negative thin space
  # Remove spaces around operators
  answer <- gsub("\\s+", "", answer)
  answer
}

#' Check a student's answer against the correct answer
#'
#' Handles equivalent forms: "3/4" = "6/8" = "0.75" = "\\frac{3}{4}".
#' Also accepts letter answers, strips parentheses, and normalizes case.
#'
#' @param problem A fisherapp_problem object (from \code{\link{generate_problem}})
#' @param student_answer Character string. The student's answer.
#' @return List with fields:
#'   \describe{
#'     \item{correct}{Logical. TRUE if the answer is correct.}
#'     \item{correct_answer}{Character. The correct answer in LaTeX format.}
#'     \item{solution_steps}{Character vector. The worked solution steps.}
#'   }
#' @export
check_answer <- function(problem, student_answer) {
  correct_raw <- problem$answer_raw
  is_correct <- compare_answers(student_answer, correct_raw)

  list(
    correct = is_correct,
    correct_answer = problem$answer,
    solution_steps = problem$solution_steps
  )
}

#' Compare a student answer to the correct answer
#'
#' Uses a multi-strategy approach:
#' 1. Letter answer matching (if answer_raw contains answer_letter)
#' 2. Numeric comparison with tolerance
#' 3. Normalized string comparison for algebraic expressions
#'
#' @param student_answer Character string
#' @param correct_raw The raw answer object from the template's solve function
#' @return Logical
#' @keywords internal
compare_answers <- function(student_answer, correct_raw) {
  student_answer <- trimws(as.character(student_answer))
  if (nchar(student_answer) == 0) return(FALSE)

  normalized_student <- normalize_answer(student_answer)

  # Strategy 1: Letter answer matching
  if (is.list(correct_raw) && !is.null(correct_raw$answer_letter)) {
    correct_letter <- normalize_answer(correct_raw$answer_letter)
    if (normalized_student == correct_letter) return(TRUE)
  }

  # Strategy 2: Numeric comparison
  parsed <- parse_student_answer(student_answer)
  if (!is.null(parsed)) {
    correct_numeric <- to_numeric(correct_raw)
    student_numeric <- to_numeric(parsed)
    if (!is.na(correct_numeric) && !is.na(student_numeric)) {
      if (isTRUE(all.equal(correct_numeric, student_numeric,
                           tolerance = 1e-9))) {
        return(TRUE)
      }
    }
  }

  # Strategy 3: Algebraic string comparison (after normalization)
  if (is.list(correct_raw) && !is.null(correct_raw$answer_expr)) {
    correct_expr <- normalize_algebra(correct_raw$answer_expr)
    student_expr <- normalize_algebra(student_answer)
    if (correct_expr == student_expr) return(TRUE)
  }

  FALSE
}

#' Normalize an algebraic expression for comparison
#'
#' Handles equivalent algebraic forms: "2*x" = "2x", "x+1" = "1+x",
#' removes unnecessary multiplication signs, and sorts commutative terms.
#'
#' @param expr Character string
#' @return Normalized character string
#' @keywords internal
normalize_algebra <- function(expr) {
  expr <- normalize_answer(expr)

  # Remove explicit multiplication signs: "2*x" -> "2x"
  expr <- gsub("\\*", "", expr)

  # Sort additive terms: split on +, sort, rejoin
  # Handle subtraction by converting "a-b" to "a+-b"
  expr_expanded <- gsub("-", "+-", expr)
  terms <- strsplit(expr_expanded, "\\+")[[1]]
  terms <- terms[nchar(terms) > 0]  # Remove empty strings
  terms <- sort(terms)
  paste(terms, collapse = "+")
}

#' Parse a student answer string into a structured form
#'
#' Supports: "3/4", "0.75", ".5", "3", "-2/5", "\\frac{3}{4}", "\\frac{-3}{4}"
#'
#' @param answer Character string
#' @return List with either num/den or value, or NULL if unparseable
#' @keywords internal
parse_student_answer <- function(answer) {
  answer <- normalize_answer(answer)

  # Try LaTeX fraction: \frac{num}{den}
  m <- regmatches(answer,
    regexec("^-?\\\\frac\\{(-?\\d+)\\}\\{(-?\\d+)\\}$", answer))[[1]]
  if (length(m) == 3) {
    num <- as.integer(m[2])
    den <- as.integer(m[3])
    # Handle leading negative sign
    if (startsWith(answer, "-") && num > 0) num <- -num
    return(list(num = num, den = den))
  }

  # Try plain fraction: num/den (also handles ".5/2" style)
  m <- regmatches(answer,
    regexec("^(-?\\d*\\.?\\d+)/(-?\\d*\\.?\\d+)$", answer))[[1]]
  if (length(m) == 3) {
    num_val <- suppressWarnings(as.numeric(m[2]))
    den_val <- suppressWarnings(as.numeric(m[3]))
    if (!is.na(num_val) && !is.na(den_val) && den_val != 0) {
      return(list(value = num_val / den_val))
    }
  }

  # Try \sqrt{n} -> evaluate to numeric
  m <- regmatches(answer,
    regexec("^\\\\sqrt\\{(-?\\d+\\.?\\d*)\\}$", answer))[[1]]
  if (length(m) == 2) {
    val <- suppressWarnings(as.numeric(m[2]))
    if (!is.na(val) && val >= 0) return(list(value = sqrt(val)))
  }

  # Try sqrt(n) -> evaluate to numeric
  m <- regmatches(answer,
    regexec("^sqrt\\((-?\\d+\\.?\\d*)\\)$", answer))[[1]]
  if (length(m) == 2) {
    val <- suppressWarnings(as.numeric(m[2]))
    if (!is.na(val) && val >= 0) return(list(value = sqrt(val)))
  }

  # Try decimal or integer (including leading dot like ".5")
  val <- suppressWarnings(as.numeric(answer))
  if (!is.na(val)) return(list(value = val))

  NULL
}

#' Convert a raw answer to numeric for comparison
#'
#' Handles various answer formats from different template types.
#'
#' @param raw Answer object (list or numeric)
#' @return Numeric value, or NA
#' @keywords internal
to_numeric <- function(raw) {
  if (is.numeric(raw) && length(raw) == 1) return(as.numeric(raw))

  if (is.list(raw)) {
    # Direct numeric value
    if (!is.null(raw$value)) return(as.numeric(raw$value))
    # Fraction from answer checker parse
    if (!is.null(raw$num) && !is.null(raw$den)) {
      return(raw$num / raw$den)
    }
    # Fraction from template solve (answer_num/answer_den convention)
    if (!is.null(raw$answer_num) && !is.null(raw$answer_den)) {
      return(raw$answer_num / raw$answer_den)
    }
    # answer_value convention (for non-fraction answers)
    if (!is.null(raw$answer_value)) return(as.numeric(raw$answer_value))
  }

  NA_real_
}

File: backend/app/services/answer_checker.py
"""
Robust answer checker using SymPy for symbolic math verification.
Handles LaTeX input from MathLive as well as plain expressions.

LaTeX conversion handles:
- \frac{a}{b}   → a/b
- x^{n}         → x**n
- \sqrt{x}      → sqrt(x)
- \log_b(x)     → log(x, b)
- \ln(x)        → log(x)
- \cdot, \times → *
- \div          → /
- \binom{n}{k}  → binomial(n,k)
- n!            → factorial(n)

Solution set handling:
- "2, 3" or "3, 2" or "x=2, x=3" or "{2, 3}" or "\{2, 3\}" all treated as sets
- Compared as sets (order-independent), partial answers marked INCORRECT
"""
import re
import logging
from typing import Literal, Optional

logger = logging.getLogger(__name__)

try:
    from sympy import (
        simplify, sympify, SympifyError, N, Symbol,
        factorial, log, sqrt, Rational, binomial,
        oo, zoo, nan
    )
    from sympy.parsing.sympy_parser import (
        parse_expr, standard_transformations, implicit_multiplication_application
    )
    SYMPY_AVAILABLE = True
except ImportError:
    SYMPY_AVAILABLE = False

AnswerType = Literal["symbolic", "numeric", "multiple_choice"]

TRANSFORMATIONS = standard_transformations + (implicit_multiplication_application,)


def latex_to_sympy_str(latex: str) -> str:
    """
    Convert a LaTeX string (as output by MathLive) to a SymPy-parseable string.
    Handles common patterns found in this app's content.
    """
    s = latex.strip()

    # Remove surrounding \( \) delimiters if present
    s = re.sub(r'^\\\(', '', s).strip()
    s = re.sub(r'\\\)$', '', s).strip()

    # Remove surrounding \[ \] delimiters if present
    s = re.sub(r'^\\\[', '', s).strip()
    s = re.sub(r'\\\]$', '', s).strip()

    # Remove surrounding $ $ delimiters if present
    s = re.sub(r'^\$+|\$+$', '', s).strip()

    # Handle LaTeX set braces \{ ... \} — strip them to expose the contents
    # e.g. \{2, 3\} → 2, 3
    s = re.sub(r'\\\{', '{', s)
    s = re.sub(r'\\\}', '}', s)

    # \frac{a}{b} → (a)/(b) — handle nested (apply multiple times)
    def replace_frac(m):
        return f'({m.group(1)})/({m.group(2)})'
    for _ in range(5):
        prev = s
        s = re.sub(
            r'\\frac\{([^{}]*(?:\{[^{}]*\}[^{}]*)*)\}\{([^{}]*(?:\{[^{}]*\}[^{}]*)*)\}',
            replace_frac, s
        )
        if s == prev:
            break

    # \sqrt[n]{x} → x**(1/n), \sqrt{x} → sqrt(x)
    s = re.sub(r'\\sqrt\[([^\]]+)\]\{([^}]+)\}', r'((\2)**(1/(\1)))', s)
    s = re.sub(r'\\sqrt\{([^}]+)\}', r'sqrt(\1)', s)

    # \log_{b}(x) → log(x, b); various forms
    # IMPORTANT: these must run BEFORE the general subscript-removal step so
    # that the base (e.g. the "2" in \log_{2}) is not stripped prematurely.
    s = re.sub(r'\\log_\{([^}]+)\}\s*\(([^)]+)\)', r'log(\2, \1)', s)
    s = re.sub(r'\\log_\{([^}]+)\}\s*([a-zA-Z0-9]+)', r'log(\2, \1)', s)
    s = re.sub(r'\\log_([0-9]+)\s*\(([^)]+)\)', r'log(\2, \1)', s)
    s = re.sub(r'\\log_([0-9]+)\s+([a-zA-Z0-9]+)', r'log(\2, \1)', s)
    s = re.sub(r'\\log\s*\(([^)]+)\)', r'log(\1, 10)', s)
    s = re.sub(r'\\log\s+([a-zA-Z0-9]+)', r'log(\1, 10)', s)
    s = re.sub(r'\\ln\s*\(([^)]+)\)', r'log(\1)', s)
    s = re.sub(r'\\ln\s+([a-zA-Z0-9]+)', r'log(\1)', s)

    # x^{n} → x**(n), x^n → x**n
    s = re.sub(r'\^\{([^}]+)\}', r'**(\1)', s)
    s = re.sub(r'\^([a-zA-Z0-9])', r'**\1', s)

    # Subscripts (indices) — remove for simple cases
    # (log subscripts have already been handled above)
    s = re.sub(r'_\{[^}]+\}', '', s)
    s = re.sub(r'_[a-zA-Z0-9]', '', s)

    # \binom{n}{k} → binomial(n, k)
    s = re.sub(r'\\binom\{([^}]+)\}\{([^}]+)\}', r'binomial(\1, \2)', s)

    # n! → factorial(n)
    s = re.sub(r'([a-zA-Z0-9]+)!', r'factorial(\1)', s)

    # Operators
    s = s.replace('\\cdot', '*').replace('\\times', '*')
    s = s.replace('\\div', '/')
    s = s.replace('÷', '/').replace('×', '*')

    # Remove grouping helpers
    s = s.replace('\\left', '').replace('\\right', '')
    s = s.replace('\\!', '').replace('\\ ', ' ')

    # Remove any remaining unknown backslash-letter commands
    s = re.sub(r'\\[a-zA-Z]+\*?', '', s)

    # Clean up extra spaces and stray braces
    s = s.replace('{', '(').replace('}', ')')
    s = re.sub(r'\s+', ' ', s).strip()

    return s


def _try_parse(expr_str: str):
    """Try multiple parsing strategies. Returns SymPy expr or None."""
    if not SYMPY_AVAILABLE:
        return None

    # Strategy 1: parse_expr with implicit multiplication
    try:
        return parse_expr(expr_str, transformations=TRANSFORMATIONS)
    except Exception:
        pass

    # Strategy 2: sympify directly
    try:
        return sympify(expr_str)
    except Exception:
        pass

    return None


def _to_sympy(raw: str):
    """
    Convert a raw answer string (LaTeX or plain text) to a SymPy expression.
    Tries LaTeX conversion first, then plain-text parsing.
    Returns the SymPy expression or None on failure.
    """
    # Try LaTeX conversion first
    converted = latex_to_sympy_str(raw)
    logger.debug(f"  latex_to_sympy_str({raw!r}) → {converted!r}")

    expr = _try_parse(converted)
    if expr is not None:
        logger.debug(f"  parsed({converted!r}) → {expr}")
        return expr

    # If LaTeX conversion changed the string, also try the original plain text
    if converted != raw:
        expr = _try_parse(raw)
        if expr is not None:
            logger.debug(f"  parsed plain({raw!r}) → {expr}")
            return expr

    logger.debug(f"  could not parse {raw!r}")
    return None


def normalize_string(s: str) -> str:
    """Basic string normalization for last-resort fallback comparison."""
    s = s.strip().lower()
    # Remove LaTeX-specific formatting
    s = re.sub(r'\\[a-zA-Z]+', '', s)
    s = s.replace('{', '').replace('}', '')
    s = s.replace(' ', '').replace('×', '*').replace('÷', '/').replace('−', '-')
    # Normalize fraction notation
    s = re.sub(r'\((\d+)\)/\((\d+)\)', r'\1/\2', s)
    return s


# ── Solution-set helpers ─────────────────────────────────────────────────────

def _is_multi_value(s: str) -> bool:
    """
    Return True if the string looks like a multi-value answer (solution set).

    A string is multi-value if it contains:
      - a comma not inside parentheses, or
      - the word "and" (as a separator), or
      - set-brace notation { ... }

    We deliberately exclude single expressions like "2x + 3" which contain
    no comma and no "and" as a separator between values.
    """
    stripped = s.strip()

    # Check for LaTeX set braces \{ \} or plain { }
    if re.search(r'[\{\}]', stripped):
        return True

    # Check for "and" used as a value separator (standalone word)
    # e.g. "x=2 and x=3"  but NOT "sin and cos" (but that won't appear here)
    if re.search(r'\band\b', stripped, re.IGNORECASE):
        return True

    # Check for a comma that is not inside parentheses
    # We scan character by character tracking parenthesis depth
    depth = 0
    for ch in stripped:
        if ch == '(':
            depth += 1
        elif ch == ')':
            depth -= 1
        elif ch == ',' and depth == 0:
            return True

    return False


def _strip_variable_prefix(token: str) -> str:
    """
    Remove leading variable assignment like "x = " or "x=" from a token.
    Examples:
      "x = 2"  → "2"
      "x=3"    → "3"
      "y = -1" → "-1"
      "2"      → "2"
    """
    # Match patterns like: letter(s) optional-spaces = optional-spaces
    m = re.match(r'^[a-zA-Z]\s*=\s*(.*)', token.strip())
    if m:
        return m.group(1).strip()
    return token.strip()


def _split_multi_value(s: str) -> list:
    """
    Split a multi-value answer string into individual value tokens.

    Handles:
      "2, 3"           → ["2", "3"]
      "x = 2, x = 3"  → ["2", "3"]
      "x=2 and x=3"   → ["2", "3"]
      "{2, 3}"         → ["2", "3"]
      "\{2, 3\}"       → ["2", "3"]
      "x = 2, 3"       → ["2", "3"]
    """
    stripped = s.strip()

    # Remove outer set braces (both LaTeX \{ \} and plain { })
    # LaTeX style: \{...\}
    stripped = re.sub(r'^\\\{(.*)\\\}$', r'\1', stripped).strip()
    # Plain style: {...}
    stripped = re.sub(r'^\{(.*)\}$', r'\1', stripped).strip()

    # Split on "and" or commas
    parts = re.split(r'\s+and\s+|,', stripped, flags=re.IGNORECASE)

    # Strip variable prefixes and whitespace from each part
    tokens = [_strip_variable_prefix(p) for p in parts]

    # Filter out empty tokens
    tokens = [t for t in tokens if t]

    return tokens


def _parse_value_set(s: str):
    """
    Parse a multi-value answer string into a frozenset of SymPy expressions.
    Returns None if any value cannot be parsed.
    """
    tokens = _split_multi_value(s)
    sympy_values = []
    for token in tokens:
        expr = _to_sympy(token)
        if expr is None:
            logger.debug(f"  _parse_value_set: could not parse token {token!r}")
            return None
        sympy_values.append(expr)
    return sympy_values  # list (will be compared as a set)


def _sympy_values_equal(a, b) -> bool:
    """
    Return True if two SymPy expressions are mathematically equal.
    Uses simplify(a - b) == 0 with numeric fallback.
    """
    try:
        diff = simplify(a - b)
        if diff == 0:
            return True
    except Exception:
        pass

    try:
        va = complex(N(a))
        vb = complex(N(b))
        if abs(va - vb) < 1e-6:
            return True
    except Exception:
        pass

    return False


def _compare_solution_sets(student_values: list, correct_values: list) -> bool:
    """
    Compare two lists of SymPy expressions as sets.
    Returns True only if the student set exactly matches the correct set
    (same cardinality, each correct value matched by exactly one student value).
    """
    if len(student_values) != len(correct_values):
        logger.info(
            f"  solution set size mismatch: "
            f"student={len(student_values)} correct={len(correct_values)}"
        )
        return False

    # Check that every correct value has a matching student value (and vice versa,
    # implied by equal size + injection).
    correct_unmatched = list(correct_values)
    for sv in student_values:
        matched = False
        for i, cv in enumerate(correct_unmatched):
            if _sympy_values_equal(sv, cv):
                correct_unmatched.pop(i)
                matched = True
                break
        if not matched:
            logger.info(f"  solution set: student value {sv} not in correct set")
            return False

    return True


# ── Main entry point ─────────────────────────────────────────────────────────

def check_answer(student_answer: str, correct_answer: str, answer_type: AnswerType = "symbolic") -> bool:
    """
    Check if a student's answer matches the correct answer.

    Detailed logging is included to help diagnose format mismatches.

    For multiple_choice: case-insensitive string match.
    For numeric: float comparison with tolerance.
    For symbolic: SymPy symbolic equivalence, with LaTeX parsing for MathLive output.
                  Also handles solution sets (multiple comma-separated values).
    """
    student = student_answer.strip()
    correct = correct_answer.strip()

    logger.info(
        f"check_answer | type={answer_type!r} | "
        f"student={student!r} | correct={correct!r}"
    )

    # ── Multiple choice ──────────────────────────────────────────────────────
    if answer_type == "multiple_choice":
        result = student.lower() == correct.lower()
        logger.info(f"  multiple_choice match: {result}")
        return result

    # ── Numeric ─────────────────────────────────────────────────────────────
    if answer_type == "numeric":
        # Direct float comparison
        try:
            sv = float(student)
            cv = float(correct)
            result = abs(sv - cv) <= 0.01
            logger.info(f"  numeric float: {sv} vs {cv} → {result}")
            return result
        except ValueError:
            pass

        # SymPy numeric evaluation
        if SYMPY_AVAILABLE:
            s_expr = _to_sympy(student)
            c_expr = _to_sympy(correct)
            if s_expr is not None and c_expr is not None:
                try:
                    sv = float(N(s_expr))
                    cv = float(N(c_expr))
                    result = abs(sv - cv) <= 0.01
                    logger.info(f"  numeric sympy: {sv} vs {cv} → {result}")
                    return result
                except Exception as e:
                    logger.debug(f"  numeric sympy N() failed: {e}")

        # Normalized string fallback
        ns = normalize_string(student)
        nc = normalize_string(correct)
        result = ns == nc
        logger.info(f"  numeric string fallback: {ns!r} vs {nc!r} → {result}")
        return result

    # ── Symbolic (default) ───────────────────────────────────────────────────
    if not SYMPY_AVAILABLE:
        ns = normalize_string(student)
        nc = normalize_string(correct)
        result = ns == nc
        logger.info(f"  symbolic (no sympy) string: {result}")
        return result

    # ── Solution-set detection ───────────────────────────────────────────────
    # If either the student answer or the correct answer looks like a multi-value
    # solution set (contains commas, "and", or set braces), handle as a set
    # comparison.  A single value against a multi-value correct answer is INCORRECT
    # (partial answers not accepted).

    student_is_multi = _is_multi_value(student)
    correct_is_multi = _is_multi_value(correct)

    if correct_is_multi or student_is_multi:
        logger.info(
            f"  solution set mode: student_multi={student_is_multi}, "
            f"correct_multi={correct_is_multi}"
        )

        # Parse both sides as value sets
        # If correct is multi but student is single, treat student as a 1-element set
        if correct_is_multi and not student_is_multi:
            # Partial answer — student only gave one value
            correct_values = _parse_value_set(correct)
            if correct_values is not None and len(correct_values) > 1:
                # More than one correct value required; single answer is wrong
                logger.info(
                    f"  solution set: student gave single value but "
                    f"{len(correct_values)} required → INCORRECT"
                )
                return False
            # Correct set has only 1 value; fall through to single comparison below

        student_values = _parse_value_set(student) if student_is_multi else None
        correct_values = _parse_value_set(correct) if correct_is_multi else None

        # If student is multi-value
        if student_values is not None:
            if correct_values is not None:
                result = _compare_solution_sets(student_values, correct_values)
                logger.info(f"  solution set comparison: {result}")
                return result
            else:
                # correct is single value; student gave multiple — need exact match of sets
                c_expr = _to_sympy(correct)
                if c_expr is not None:
                    # Student must have given exactly one value equal to correct
                    if len(student_values) == 1 and _sympy_values_equal(student_values[0], c_expr):
                        logger.info("  solution set: student multi (1 val) == correct single")
                        return True
                logger.info("  solution set: student multi != correct single")
                return False

        # correct_is_multi but student is single — we already handled the >1 case above
        # Here correct_values has exactly 1 element
        if correct_values is not None:
            s_expr = _to_sympy(student)
            if s_expr is not None and len(correct_values) == 1:
                result = _sympy_values_equal(s_expr, correct_values[0])
                logger.info(f"  solution set (single correct): {result}")
                return result

        # Fallback: normalized string comparison for sets
        ns = normalize_string(student)
        nc = normalize_string(correct)
        result = ns == nc
        logger.info(f"  solution set string fallback: {ns!r} vs {nc!r} → {result}")
        return result

    # ── Standard single-value symbolic comparison ────────────────────────────
    s_expr = _to_sympy(student)
    c_expr = _to_sympy(correct)

    if s_expr is not None and c_expr is not None:
        # Symbolic simplification
        try:
            diff = simplify(s_expr - c_expr)
            if diff == 0:
                logger.info("  symbolic simplify: match (diff=0)")
                return True
        except Exception as e:
            logger.debug(f"  simplify failed: {e}")

        # Numeric evaluation backup
        try:
            sv = complex(N(s_expr))
            cv = complex(N(c_expr))
            if abs(sv - cv) < 1e-6:
                logger.info(f"  symbolic numeric eval: match ({sv} ≈ {cv})")
                return True
        except Exception as e:
            logger.debug(f"  numeric eval failed: {e}")

        logger.info(f"  symbolic: no match ({s_expr!r} vs {c_expr!r})")
    else:
        logger.info(f"  symbolic: parse failed (s={s_expr}, c={c_expr})")

    # Normalized string fallback
    ns = normalize_string(student)
    nc = normalize_string(correct)
    result = ns == nc
    logger.info(f"  symbolic string fallback: {ns!r} vs {nc!r} → {result}")
    return result
